### PR TITLE
[Profiling]  Update elementwise op

### PR DESCRIPTION
**Pull request by chengduoZH**
Fix https://github.com/PaddlePaddle/Paddle/issues/6166

"We found the performance of `Eigen::Tensor::broadcast` is not good if the `Tensor` is not just a scalar, because the `Eigen::Tensor::broadcast` try to implement a very general `broadcast` method. However, the most of `broadcast` operators in the neural network is `rowwise` or `colwise`. The implementation of `rowwise` or `colwise` is much simpler and faster than a general implementation.
The `Eigen::Tensor::broadcast` is mainly implmeneted by this [function](https://github.com/RLovelett/eigen/blob/master/unsupported/Eigen/CXX11/src/Tensor/TensorBroadcasting.h#L172 )." —— @reyoung

### CODE REVIEW

File: paddle/operators/elementwise_op_function.h [hunk 1]
```diff
@@ -54,6 +59,147 @@ inline void get_mid_dims(const framework::DDim& x_dims,
...
+  int64_t n_;
+};
+
+template <typename T>
+struct MidWiseTransformIterator<T, platform::CPUPlace> {
+  MidWiseTransformIterator(const T* ptr, int n, int post)
+      : ptr_(ptr), i_(0), j_(0), n_(n), post_(post) {}
+
+  MidWiseTransformIterator<T, platform::CPUPlace>& operator++() {
+    i_ = ++j_ / post_ % n_;
```
Comment (reyoung) [inline paddle/operators/elementwise_op_function.h]: Please add parentheses here.

File: paddle/operators/elementwise_op_function.h [hunk 2]
```diff
@@ -54,6 +59,147 @@ inline void get_mid_dims(const framework::DDim& x_dims,
...
+  }
+
+  bool operator==(
+      const MidWiseTransformIterator<T, platform::CPUPlace>& rhs) const {
+    return (ptr_ + i_) == &(*rhs);
+  }
+
+  bool operator!=(
+      const MidWiseTransformIterator<T, platform::CPUPlace>& rhs) const {
+    return (ptr_ + i_) &= &(*rhs);
```
Comment (reyoung) [inline paddle/operators/elementwise_op_function.h]: `&=` -> `!=`

File: paddle/operators/elementwise_op_function.h [hunk 3]
```diff
@@ -54,6 +59,147 @@ inline void get_mid_dims(const framework::DDim& x_dims,
...
+#ifdef __NVCC__
+template <typename T>
+struct RowwiseTransformIterator<T, platform::GPUPlace>
+    : public thrust::iterator_adaptor<
+          RowwiseTransformIterator<T, platform::GPUPlace>, const T*> {
+ public:
+  typedef thrust::iterator_adaptor<
+      RowwiseTransformIterator<T, platform::GPUPlace>, const T*>
+      super_t;
+  __host__ __device__ RowwiseTransformIterator(const T* x, int n)
```
Comment (reyoung) [inline paddle/operators/elementwise_op_function.h]: Please use `HOSTDEVICE` macro

File: paddle/operators/elementwise_op_function.h [hunk 4]
```diff
@@ -54,6 +59,147 @@ inline void get_mid_dims(const framework::DDim& x_dims,
...
+      : super_t(x), begin_(x), n_(n){};
+  friend class thrust::iterator_core_access;
+
+ private:
+  unsigned int n_;
+  const T* begin_;
+  __host__ __device__ typename super_t::reference dereference() const {
```
Comment (reyoung) [inline paddle/operators/elementwise_op_function.h]: Please use `HOSTDEVICE` macro

File: paddle/operators/elementwise_add_op.h [hunk 5]
```diff
@@ -19,11 +19,48 @@
...
+struct AddFunctor {
+  HOSTDEVICE T operator()(T a, T b) const { return a + b; }
+};
+
 template <typename Place, typename T>
 class ElementwiseAddKernel : public framework::OpKernel<T> {
  public:
   void Compute(const framework::ExecutionContext& ctx) const override {
-    ElementwiseCompute<EigenAddFunctor, Place, T>(ctx);
+    using Tensor = framework::Tensor;
```
Comment (reyoung) [inline paddle/operators/elementwise_add_op.h]: Maybe we should implement all elemwise operators based on this method.

Not in hurry, but need an issue to record the following works.

Comment (chengduoZH) [reply paddle/operators/elementwise_add_op.h]: We can do these in next PR.

File: paddle/operators/elementwise_op_function.h (hunk 2 above)
Comment (chengduoZH) [reply paddle/operators/elementwise_op_function.h]: Done

File: paddle/operators/elementwise_op_function.h [hunk 6]
```diff
@@ -54,6 +59,147 @@ inline void get_mid_dims(const framework::DDim& x_dims,
...
+    return *(begin_ + (((this->base() - begin_) / post_) % n_));
+  }
+};
+#endif
+
+template <typename Functor, typename T, typename Place>
+struct TransformFunctor {
+  TransformFunctor(const framework::Tensor* x, const framework::Tensor* y,
+                   framework::Tensor* z, const framework::ExecutionContext& ctx,
```
Comment (QiJune) [inline paddle/operators/elementwise_op_function.h]: Better to use DeviceContext other than ExecutionContext

File: paddle/operators/elementwise_op_function.h (hunk 3 above)
Comment (chengduoZH) [reply paddle/operators/elementwise_op_function.h]: Done

File: paddle/operators/elementwise_op_function.h (hunk 4 above)
Comment (chengduoZH) [reply paddle/operators/elementwise_op_function.h]: Done

File: paddle/operators/elementwise_op_function.h (hunk 1 above)
Comment (chengduoZH) [reply paddle/operators/elementwise_op_function.h]: Done

File: paddle/operators/elementwise_op_function.h (hunk 6 above)
Comment (chengduoZH) [reply paddle/operators/elementwise_op_function.h]: Maybe we will replace `ElementwiseCompute` with `ElementwiseWithIteratorCompute` in next PR. So I write this for a while.

Comment (chengduoZH) [reply paddle/operators/elementwise_op_function.h]: Done

File: paddle/operators/elementwise_op_function.h [hunk 7]
```diff
@@ -54,6 +59,147 @@ inline void get_mid_dims(const framework::DDim& x_dims,
...
+    trans(ctx_, x_, x_ + nx_, MidWiseTransformIterator<T, Place>(y_, n, post),
+          z_, func_);
+  }
+
+  const T* x_;
+  const T* y_;
+  T* z_;
+  int64_t nx_;
+  const platform::DeviceContext& ctx_;
+  Functor func_;
```
Comment (qingqing01) [inline paddle/operators/elementwise_op_function.h]: The naming rules of data members for `struct` is different from `class`. 

https://google.github.io/styleguide/cppguide.html#Variable_Names

Comment (chengduoZH) [reply paddle/operators/elementwise_op_function.h]: Done, I have replaced `struct` with `class`.

**Pull request merged**